rust/src/graph/higher_order_queries.rs

```diff
@@ -0,0 +1,210 @@
+
+        self.generate_module_permutations(levels)
+            .into_par_iter()
+            .try_fold(
+                Vec::new,
+                |mut v: Vec<PackageDependency>, (from_package, to_package)| -> GrimpResult<_> {
+                    if let Some(dep) = self.find_illegal_dependencies(
+                        from_package,
+                        to_package,
+                        &all_layer_modules,
+                    )? {
+                        v.push(dep);
+                    }
+                    Ok(v)
+                },
+            )
+            .try_reduce(
+                Vec::new,
+                |mut v: Vec<PackageDependency>, package_dependencies| {
+                    v.extend(package_dependencies);
+                    Ok(v)
+                },
+            )
+    }
+
+    fn generate_module_permutations(&self, levels: &[Level]) -> Vec<(ModuleToken, ModuleToken)> {
+        let mut permutations = vec![];
+
+        for (index, level) in levels.iter().enumerate() {
+            for module in &level.layers {
+                // Should not be imported by lower layers.
+                for lower_level in &levels[index + 1..] {
+                    for lower_module in &lower_level.layers {
+                        permutations.push((*lower_module, *module));
+                    }
+                }
+
+                // Should not import siblings (if level is independent)
+                if level.independent {
+                    for sibling_module in &level.layers {
+                        if sibling_module == module {
+                            continue;
+                        }
+                        permutations.push((*module, *sibling_module));
+                    }
+                }
+            }
+        }
+
+        permutations
+    }
+
+    fn find_illegal_dependencies(
+        &self,
+        from_layer: ModuleToken,
+        to_layer: ModuleToken,
+        all_layers_modules: &FxHashSet<ModuleToken>,
+    ) -> GrimpResult<Option<PackageDependency>> {
+        // Shortcut the detailed implementation in the case of no chains.
+        // This will be much faster!
+        if !self.chain_exists(from_layer, to_layer, true)? {
+            return Ok(None);
+        }
+
+        let from_layer_with_descendants = from_layer.conv::<FxHashSet<_>>().with_descendants(self);
+        let to_layer_with_descendants = to_layer.conv::<FxHashSet<_>>().with_descendants(self);
+
+        // Disallow chains via other layers.
+        let excluded_modules =
+            all_layers_modules - &(&from_layer_with_descendants | &to_layer_with_descendants);
+
+        // Disallow chains via these imports.
+        // We'll add chains to this set as we discover them.
+        let mut excluded_imports = FxHashMap::default();
+
+        // Collect direct imports...
+        let mut direct_imports = vec![];
+        // ...and the middles of any indirect imports.
+        let mut middles = vec![];
+        loop {
```

Comment:
How confident are we that this loop will complete? I don't think there are any tests for handling cycles in `find_illegal_dependencies_for_layers`, we should probably add some.

Reply:
We've chatted - I'm more confident now.